### PR TITLE
feat: gaussian naive bayes classifier

### PR DESCRIPTION
**Pull request by Lewuathe**
Add Gaussian naive bayes classifier.

See: https://github.com/javascriptdata/scikit.js/issues/42

### CODE REVIEW

File: src/naive_bayes/baseNaiveBayes.ts [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+  varSmoothing?: number
+}
+
+export abstract class BaseNaiveBayes {
+  priors?: tf.Tensor1D
+  varSmoothing: number
+
+  private classes: tf.Tensor1D
+  private means: tf.Tensor1D[]
+  private variances: tf.Tensor1D[]
```
Comment (dcrescim) [inline src/naive_bayes/baseNaiveBayes.ts:40]: I think these should be public. If someone runs
```
let nb = new GaussianNB()
await nb.fit(X,y)
console.log(nb.classes) // would generate Typescript error
```

File: src/naive_bayes/baseNaiveBayes.ts [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+
+  constructor(params: NaiveBayesParams) {
```
Comment (dcrescim) [inline src/naive_bayes/baseNaiveBayes.ts:42]: `constructor(params: NaiveBayesParams = {})`

That way someone doesn't need to pass an empty object. If you don't have the default object then the following code errors

```
let nb = GaussianNB() // would error without default arg
```

File: src/naive_bayes/baseNaiveBayes.ts [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        const numInstances = tf.sum(mask)
+        const mean = tf.mul(features, mask.expandDims(1))
+          .sum(0)
+          .div(numInstances)
+        const variance = tf.sub(features, mean)
+          .mul(mask.expandDims(1))
+          .pow(2)
+          .sum(0)
+          .div(numInstances)
+          .add(epsilon)
```
Comment (dcrescim) [inline src/naive_bayes/baseNaiveBayes.ts:84]: I love this chunk of math @Lewuathe. It's elegant and beautiful 😍

File: src/naive_bayes/baseNaiveBayes.ts [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+    const features = convertToNumericTensor2D(X)
+    const labels = convertToTensor1D(y)
+
+    const {values, meansByLabel, variancesByLabel} = tf.tidy(() => {
+      polyfillUnique(tf)
+      const meansByLabel: Tensor1D[] = []
+      const variancesByLabel: Tensor1D[] = []
+
+      // Get the list of unique labels
+      const {values, indices} = tf.unique(labels)
```
Comment (dcrescim) [inline src/naive_bayes/baseNaiveBayes.ts:68]: `const {values} = tf.unique(labels)`

We don't use indices, right?

File: src/naive_bayes/baseNaiveBayes.ts [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+      const {mean, variance} = tf.moments(features, 0)
```
Comment (dcrescim) [inline src/naive_bayes/baseNaiveBayes.ts:70]: `const { variance } = tf.moments(features, 0)`
We don't use mean either. My vote is to just get rid of them, and save the computer some work.

File: src/naive_bayes/baseNaiveBayes.ts [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+    return probs.argMax(1)
+  }
+
+  /**
+   * Kernel function to calculate posterior probability, which should be implemented in the subclass.
+   * @param features
+   * @param mean
+   * @param variance
+   */
+  protected abstract kernel(features: tf.Tensor2D, mean: tf.Tensor1D, variance: tf.Tensor1D): tf.Tensor1D
```
Comment (dcrescim) [inline src/naive_bayes/baseNaiveBayes.ts:148]: I love the abstract method here, @Lewuathe

File: src/naive_bayes/gaussianNaiveBayes.test.ts [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+* ==========================================================================
+*/
+import { GaussianNB } from './gaussianNaiveBayes'
+
+describe.only('GaussianNB', function () {
```
Comment (dcrescim) [inline src/naive_bayes/gaussianNaiveBayes.test.ts:17]: `describe.only` -> `describe`

I'm pretty sure we don't need the `only` here.

**Review by dcrescim — APPROVED** [review]
@Lewuathe This is awesome work! Nice job! I went through and added a couple comments here and there, but this is ready for primetime.

Aside from the comments in the code, I think we need to do 2 last things before we merge.

1. Run `npx prettier --write src` to run prettier over the codebase 
2. Add `GaussianNB` to the index.ts

Other than those and the small tweaks below, I think we are cooking with fire 🔥

File: src/naive_bayes/baseNaiveBayes.ts (hunk 4 above)
Comment (Lewuathe) [reply src/naive_bayes/baseNaiveBayes.ts:68]: Oh, you know the better format. Thanks!

File: src/naive_bayes/gaussianNaiveBayes.test.ts (hunk 7 above)
Comment (Lewuathe) [reply src/naive_bayes/gaussianNaiveBayes.test.ts:17]: True

**Pull request merged**